### PR TITLE
Add writting and array modifying features to Lethe_pyvista_tools

### PR DESCRIPTION
**Pull request by voferreira**
# Description of the problem

- We wanted to create a post-processing tool to create or modify an array according to a given condition and resulting values.
- We also wanted to be able to create new vtu and pvd files so that it was possible to see the modifications on Paraview

# Description of the solution

- Add modifier to post-processing module: The modifier takes a string with a condition for the position ("x", "y", "z"), velocity ("u", "v", "w"), time ("t"), or any other array and modifies (or creates) an array according to the condition;
- Another argument takes the resulting value, modifying function as a string, or a numpy array, and uses it to modify or create the array.
- A writing function was also implemented.

# How Has This Been Tested?

- [x] Liquid fluidized bed
- [x] Hopper
- [x] small rotating drum

# Documentation

- [x] Add example file.
- [x] Add application example

# Future changes

- Examples will be added to the documentation in the near future.

### CODE REVIEW

File: contrib/postprocessing/Example.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+#To use the Lethe_pyvista_tools, you need to have python 3
```
Comment (blaisb) [inline contrib/postprocessing/Example.py:1]: please avoid capital letters in file names. example.py would be better.

File: contrib/postprocessing/Example.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+#The modules necessary to run Lethe pyvista tools are:
+#Pandas: pip install pandas
+#PyVista: pip install pyvista
+
+#To use Lethe_pyvista_tools, put the "Lethe_pyvista_tools.py" file inside
+#The same directory as your python script and procceed as follows
+
+#This line imports all Lethe_pyvista_tools functionalities
+from Lethe_pyvista_tools import *
```
Comment (blaisb) [inline contrib/postprocessing/Example.py:12]: ideally it should be lethe_pyvista_tools

File: examples/multiphysics/3d-dam-break/Lethe_pyvista_tools.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
+#############################################################################
+"""Class of methods to post-process lethe results with pyvista"""
```
Comment (blaisb) [inline examples/multiphysics/3d-dam-break/Lethe_pyvista_tools.py:2]: I am not sure I understand. Is it an alternative code to post-process the dam break case? This is not very clear to me . Can you explain a bit more what this is doing and why this has been added to this example?

**Review by blaisb — CHANGES_REQUESTED** [review]
I am not sure I fully understand some of the changes. In addition, I'm not sure about the things added to the multiphysics example.
Please avoid capital letters in file name, this is a pain to manage and it,s not really pythonic :)

File: examples/multiphysics/3d-dam-break/Lethe_pyvista_tools.py (hunk 3 above)
Comment (voferreira) [reply examples/multiphysics/3d-dam-break/Lethe_pyvista_tools.py:2]: Sorry, this file shouldn't be here. It stayed here when I switched branches. It was a mistake and I just deleted it.

**Comment by voferreira** [comment]
@blaisb It is not. I am adding one example to the documentation to make it clearer. I may not take long to finish it.

**Review by acdaigneault — COMMENTED** [review]
Nice feature! This is my first batch of comments. I still have some issues to run it right now on a other case, but I saw you did some modifications, so I want to give my current comments.

File: doc/source/examples/dem/small-scale-rotating-drum-post-processing/small-scale-rotating-drum-post-processing.rst [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+Description of the case
+-----------------------
+
+In this example, we illustrate the mixing inside a rotating drum by coloring the particles at the first time-step according to their radial position. To do so, we post-process Lethe-DEM data using `Python <https://www.python.org/>`_, `PyVista <https://docs.pyvista.org/>`_, `lethe_pyvista_tools <https://github.com/lethe-cfd/lethe/tree/master/contrib/postprocessing>`_, and `ParaView <https://www.paraview.org/>`_
+
+The DEM files used in this example are obtained following the `Small scale rotating drum example <../../dem/rotating-drum/small-scale-rotating-drum.html>`_.
+
+.. note::
+  It is not necessary to use all mentioned tools, but they are used in this example to show different possibilities of data access according to user's need.
```
Comment (blaisb) [inline doc/source/examples/dem/small-scale-rotating-drum-post-processing/small-scale-rotating-drum-post-processing.rst:38]:
```suggestion
  It is not necessary to use all mentioned tools, but they are used in this example to show different ways to process the data according to user's need.
```

File: examples/dem/small-scale-rotating-drum-postprocessing/packing-rotating-drum.prm [hunk 5]
```diff
@@ -0,0 +1,110 @@
+# Listing of Parameters
```
Comment (blaisb) [inline examples/dem/small-scale-rotating-drum-postprocessing/packing-rotating-drum.prm:1]: Why do you need to make a copy of those PRM files here?

**Review by blaisb — APPROVED** [review]
Very minor comments, very very good work :)!

**Comment by blaisb** [comment]
@voferreira On my side this is good. Just ping me when you are ready for "Le merge"

File: doc/source/examples/dem/small-scale-rotating-drum-post-processing/small-scale-rotating-drum-post-processing.rst [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+- Post-processing using `Python <https://www.python.org/>`_, `PyVista <https://docs.pyvista.org/>`_, `lethe_pyvista_tools <https://github.com/lethe-cfd/lethe/tree/master/contrib/postprocessing>`_, and `ParaView <https://www.paraview.org/>`_.
+
+
+Files used in this example
+----------------------------
+
+- Parameters file for particle insertion: ``/examples/postprocessing/small-scale-rotating-drum-postprocessing/packing-rotating-drum.prm``
+- Parameters file for drum rotation: ``/examples/postprocessing/small-scale-rotating-drum-postprocessing/small-rotating-drum-dem.prm``
+- Python module for Lethe data post-processing: ``/contrib/postprocessing/lethe_pyvista_tools.py``
+- Python script using module for rotating drum post-processing: ``/examples/postprocessing/small-scale-rotating-drum-postprocessing/example_small_rotating_drum.py``
```
Comment (acdaigneault) [inline doc/source/examples/dem/small-scale-rotating-drum-post-processing/small-scale-rotating-drum-post-processing.rst:26]: Why ``/examples/postprocessing/small-scale-rotating-drum-postprocessing/example_small_rotating_drum.py``?
Is not the example in the examples/dem repository?

File: examples/dem/small-scale-rotating-drum-postprocessing/example_small_rotating_drum.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+#Read files and store data to object named "particles"
+particles.read_lethe_to_pyvista("out.pvd")
+
+
+#State condition for particle_color array creation
+condition = "(y**2 + z**2)**(1/2) > 0.025"
+
+#Create array named particle_color and give the value of 1 to all particles within the given condition
+#The packing of particles is finished after 40th time step
```
Comment (acdaigneault) [inline examples/dem/small-scale-rotating-drum-postprocessing/example_small_rotating_drum.py:25]: Is it really the 40th time step? Is it not the dt*frequency of output? Or the 40th output file. 
Maybe "reference_time_step" is a bit unclear. What do you think?

File: doc/source/examples/dem/small-scale-rotating-drum-post-processing/small-scale-rotating-drum-post-processing.rst [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+  
+  If we set ``first = 2``, ``particles.df[0]`` corresponds to the dataset of the 2nd time-step of the simulation.
+
+
+Creation of a new array
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+To color the particles according to their position, we use the function ``array_modifier``, which takes the following arguments:
+
+- ``reference_array_name``: Name of the array used to sort the data. By default: ``"ID"``
```
Comment (acdaigneault) [inline doc/source/examples/dem/small-scale-rotating-drum-post-processing/small-scale-rotating-drum-post-processing.rst:103]: I know that saying "array used to sort the data" and "ID" means you use the ID of particle for sorting, but what else can it be? If there's other nice option, how can it be relevant?

File: doc/source/examples/dem/small-scale-rotating-drum-post-processing/small-scale-rotating-drum-post-processing.rst [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+- ``restart_array``: If ``True``, zeroes the entire array before modifying it. If one wants to modify part of the array keeping the rest intact, it must be set as ``False``. By default: ``False``
+- ``condition``: Takes a string and uses it in an if condition to modify the array. Variables accepted include ``x``, ``y``, ``z``, ``u``, ``v``, ``w``, ``t``, and any other array (``ID`` for example). It also accepts a combination of them, such as ``"x*w**2 + t > 2 and ID > 0"``. By default: ``""``
+- ``array_values``: New values to the array. This argument accepts a single value (which will be repeated to all data respecting the given ``condition``), a `NumPy <https://numpy.org/>`_ array, or `Python list <https://docs.python.org/3/tutorial/datastructures.html>`_ (with the same length (``len``) of all other arrays in ``particles.df``), or a string such as ``"2*x + t"`` (working just like the ``condition`` argument). By default: ``0``
+- ``standard_value``: If ``restart array = True`` or the array is a completely new array, the ``standard_value`` will be plugged to the entire array before modifying it. By default: ``0``
```
Comment (acdaigneault) [inline doc/source/examples/dem/small-scale-rotating-drum-post-processing/small-scale-rotating-drum-post-processing.rst:108]: I think the description of the line 105 (restart_array) should not says "zeroes the entire array before modifying it" if you can choose this value with this parameters, even if 0 is the default value :)

File: doc/source/examples/dem/small-scale-rotating-drum-post-processing/small-scale-rotating-drum-post-processing.rst [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+- ``time_dependent``: ``array_modifier`` can be time dependent or not. If set ``True``, ``condition`` will be tested to each of the time-steps, while if it is ``False``, it will be applied using the ``reference_time_step`` instead, and the modification will be just replicated to the other time-steps. By default: ``False``
+- ``reference_time_step``: Reference time-step to which the modification will be applied. If ``time_dependent = False``, the result of the modificaition applied to ``reference_time_step`` will be simply replicated to the others. By default: ``0``
```
Comment (acdaigneault) [inline doc/source/examples/dem/small-scale-rotating-drum-post-processing/small-scale-rotating-drum-post-processing.rst:110]: This parameter is so powerful!! :D 
As already said in this PR, I think you should explain that this is not really the time step, but the output step.

File: doc/source/examples/dem/small-scale-rotating-drum-post-processing/small-scale-rotating-drum-post-processing.rst [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+  <iframe width="560" height="315" src="https://www.youtube.com/embed/qxO4MD_zg2w" title="Rotating drum - mixing study" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>
+
+The simulation results show that the layers in red and green, that is, particles close to the walls, tend to mix faster than the ones in the center.
+
+Possibilities for extension
+----------------------------
+
+- Give a different ``condition`` to create the ``particle_color`` array
+- Use the ``lethe_pyvista_tools`` for a different problem, modifying the ``condition`` accordingly
+- Use the tools in the `PyVista oficial repository <https://docs.pyvista.org>`_ to create screenshots, movies, and plots with the data.
```
Comment (acdaigneault) [inline doc/source/examples/dem/small-scale-rotating-drum-post-processing/small-scale-rotating-drum-post-processing.rst:212]:
```suggestion
- Use the tools in the `PyVista official repository <https://docs.pyvista.org>`_ to create screenshots, movies, and plots with the data.
```

**Review by acdaigneault — COMMENTED** [review]
Wonderful tool and very nice and concise documentation! Those are my comments about the code, I will test it later and will add some comment regards that if needed. Super good job!

**Review by acdaigneault — APPROVED** [review]

**Pull request merged**